storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java

```diff
@@ -0,0 +1,873 @@
+        this.lazyOffsetIndex = lazyOffsetIndex;
+        this.lazyTimeIndex = lazyTimeIndex;
+        this.txnIndex = txnIndex;
+        this.baseOffset = baseOffset;
+        this.indexIntervalBytes = indexIntervalBytes;
+        this.rollJitterMs = rollJitterMs;
+        this.time = time;
+        this.created = time.milliseconds();
+    }
+
+    // Visible for testing
+    public LogSegment(LogSegment segment) {
+        this(segment.log, segment.lazyOffsetIndex, segment.lazyTimeIndex, segment.txnIndex, segment.baseOffset,
+                segment.indexIntervalBytes, segment.rollJitterMs, segment.time);
+    }
+
+    public OffsetIndex offsetIndex() throws IOException {
+        return lazyOffsetIndex.get();
+    }
+
+    public File offsetIndexFile() {
+        return lazyOffsetIndex.file();
+    }
+
+    public TimeIndex timeIndex() throws IOException {
+        return lazyTimeIndex.get();
+    }
+
+    public File timeIndexFile() {
+        return lazyTimeIndex.file();
+    }
+
+    public long baseOffset() {
+        return baseOffset;
+    }
+
+    public FileRecords log() {
+        return log;
+    }
+
+    public long rollJitterMs() {
+        return rollJitterMs;
+    }
+
+    public TransactionIndex txnIndex() {
+        return txnIndex;
+    }
+
+    public boolean shouldRoll(RollParams rollParams) throws IOException {
+        boolean reachedRollMs = timeWaitedForRoll(rollParams.now, rollParams.maxTimestampInMessages) > rollParams.maxSegmentMs - rollJitterMs;
+        int size = size();
+        return size > rollParams.maxSegmentBytes - rollParams.messagesSize ||
+            (size > 0 && reachedRollMs) ||
+            offsetIndex().isFull() || timeIndex().isFull() || !canConvertToRelativeOffset(rollParams.maxOffsetInMessages);
+    }
+
+    public void resizeIndexes(int size) throws IOException {
+        offsetIndex().resize(size);
+        timeIndex().resize(size);
+    }
+
+    public void sanityCheck(boolean timeIndexFileNewlyCreated) throws IOException {
+        if (offsetIndexFile().exists()) {
+            // Resize the time index file to 0 if it is newly created.
+            if (timeIndexFileNewlyCreated)
+              timeIndex().resize(0);
+            // Sanity checks for time index and offset index are skipped because
+            // we will recover the segments above the recovery point in recoverLog()
+            // in any case so sanity checking them here is redundant.
+            txnIndex.sanityCheck();
+        } else
+            throw new NoSuchFileException("Offset index file " + offsetIndexFile().getAbsolutePath() + " does not exist");
+    }
+
+    public TimestampOffset maxTimestampAndOffsetSoFar() throws IOException {
+        if (maxTimestampAndOffsetSoFar == TimestampOffset.UNKNOWN)
+            maxTimestampAndOffsetSoFar = timeIndex().lastEntry();
+        return maxTimestampAndOffsetSoFar;
+    }
+
+    /* The maximum timestamp we see so far */
+    public long maxTimestampSoFar() {
```

Comment:
could be default access

Reply:
It cannot since `UnifiedLog` is in a different package right now.

Reply:
Could we check for functions used only by `LogSegments`? I left a comment for few of them as examples. I am insisting on this because having a tighter/strict contract even between internal code reduces probability of bugs since implementation (and other details such as lifecycle) remains encapsulated to a component. Then, other callers can call public functions without having to worry about gotchas or unexpected side-effects.

Reply:
I did. There were exactly 3 cases, the ones you pointed out. You don't need to sell me on the benefit, I filed KAFKA-14489 from the start of this effort. ;)

I originally intended to do the `LogSegments` conversion in a separate PR and the 3 examples couldn't be default package at first. Move once I added the `LogSegments` conversion to this PR, it opened up that opportunity.

Reply:
> I filed [KAFKA-14489](https://issues.apache.org/jira/browse/KAFKA-14489) from the start of this effort. 

Nice! I thought my comment needed explanation because you pushed back on the index implementation leak discussion in the other comment. Glad to see that we are on same page on restricting access amongst components (as long as it doesn't impact perf such as index case).